### ISSUE TITLE
XS worker-based agoric chain cannot resume after reboot: Allocate meter exceeded

### ISSUE THREAD

**Issue by warner** [Agoric/agoric-sdk#2865]
## Describe the bug

Something funny is going on when restarting a chain under the XS workers. The chain starts up fine the first time around, but then when I kill the process and restart it, I get a metering fault during vat replay. From the 'slog' logs, it looks like vatID=1 (the "board" vat) replays fine, but vatID=11 (the "bootstrap" vat) fails.

## To Reproduce

* follow the https://agoric.com/documentation/guides/agoric-cli/starting-multiuser-dapps.html#example docs far enough to start the chain, except use `SWINGSET_WORKER_TYPE=xs-worker agoric start local-chain`
* let that run for a few minutes to initialize all vats, until cosmos-sdk starts printing "executed block NN" messages (these don't happen while swingset is starting)
* kill the process
* run `SWINGSET_WORKER_TYPE=xs-worker agoric start local-chain` again
* wait a minute or two for swingset to finish reloading the vats, whereupon it starts trying to replay their transcripts
* observe the chain process die with the following error:

```
Cannot initialize Controller RangeError: Allocate meter exceeded
```

This is running on a private branch, but I believe the code is substantially equivalent to commit af4ca46b4efd4eab04c8dd4267cb76de629eaecb, which is slightly before the `agoric@0.12.8` tag, and I think the XS code is basically the same.

## Theories

It's really curious that the replay fails, but not the original startup invocation. I tried modifying `xsnap.js` to increase the metering limit (from the default 1e7 to the maximum `2**32`-epsilon) without results.

My hunch is that something isn't initialized or prepared or reset for transcript replay the same way it gets set for regular deliveries. The xsnap side ought to be identical for both times, and I don't think we're sending a new metering limit to the worker with each delivery, so I can't imagine what could be different. Another possibility is that the results we're passing back for the simulated syscalls are not correct, and triggering more memory allocation on the worker than the first time around.

I don't know what, exactly, the XS metering limit is using as a denomimation. It's possible that the memory portion of it is in bytes, in which case 1e7 is clearly too small, but A: it worked the first time through, and B: `2**32` wasn't enough either.

**Comment by dckc** [comment]
I'm able to reproduce the problem on master at e3820735d .

**Comment by dckc** [comment]
@warner I'm trying to use the xsnap test record & play for this. The results of recording are:

[xsnap-test-2865.zip](https://github.com/Agoric/agoric-sdk/files/6313706/xsnap-test-2865.zip)

but when I play them back, I get `eval/9.js (72) # Error: buffer overflow!`.

It looks like the parser buffer size in the xsnap.c that does replay isn't as big as the one we're using to record. :-/

Increasing the size is straightforward, if tedious.

@phoddie another surprising thing: though the recorded files go up to 64, play back only shows 31 before stopping. I'm not sure why it's stopping. The process exit code is 0 as if it succeeded.

**Comment by dckc** [comment]
a little progress today; I got a stack trace... though I'm struggling to understand it.

@kriskowal @michaelfig any idea how we got from `netstring.js` to `tame.js`?

```
2021-04-21T20:12:17.850Z SwingSet: kernel: @@meter! (RangeError#1)
RangeError#1: Allocate meter exceeded

  at RangeError (packages/tame-metering/src/tame.js:131:18)
  at Function.eval [as a] (packages/transform-metering/src/meter.js:129:21)
  at meteredConstructor.copyWithin (packages/tame-metering/src/tame.js:206:41)
  at Proxy.reader (packages/xsnap/src/netstring.js:58:18)
  at reader.next (<anonymous>)
  at reader.next (packages/tame-metering/src/tame.js:202:19)
  at runToIdle (packages/xsnap/src/xsnap.js:152:64)

Cannot initialize Controller RangeError: Allocate meter exceeded
```

**Comment by kriskowal** [comment]
Looks to me like the meter ran out while receiving a message from the
parent. It says its an allocate meter exhaustion but inside copyWithin it
could just as likely be compute exhaustion. The netstring reader
occasionally has to grow its buffer for large messages. It doubles until
big enough.

**Comment by dckc** [comment]
Right... what confused me is that netstring.js is running in the kernel (as part of the vat manager) so it shouldn't be metered.

Metering is getting "stuck" on after the 1st dynamic vat. @michaelfig and I found a proof-of-concept fix: 60478bf71

![kernel-metered](https://user-images.githubusercontent.com/150986/116687047-3679fa80-a97a-11eb-9f2a-b1cc532830ba.png)

**Comment by warner** [comment]
Huh, I wonder if the kernel-side global metering didn't get turned off properly from the previous crank (on some manager=local dynamic-therefore-metered vat). This could be another consequence of #2908, if the sequence looked like:

* start creation of XS vat
  * stuff begins happening in the background
  * erroneously think the crank is done
* begin crank of a metered dynamic local vat
  * local supervisor calls `setImmediate` to schedule end-of-crank meter shutdown
  * local supervisor launches vat code
  * XS response arrives, gets put on IO queue
  * dynamic vat cranks finishes
  * if the IO queue takes priority over setImmediate:
    * XS response handler runs, but meter is still active
    * the netstring parser is not instrumented, but global objects are, and they see an active meter
    * if allocate usage in netstring parsing, plus allocation performed by the dynamic-vat crank that still hasn't finished, is greater than the dynamic-vat crank's allowed limit, then netstring parsing will fail with allocation error

I'd be surprised if an inbound IO action takes priority over setImmediate, but maybe there's something similar going on.

If that's the case, #2928 should fix it.

**Comment by warner** [comment]
It might be worthwhile to leave this flag in place, and check it at the beginning of any handler that reacts to external input. We have a flag sort of like that in `kernel.js` (`processQueueRunning`), but it's meant to detect reentrancy, whereas this new flag is about protecting one part of execution (receiving child-process messages) from a different part (local vats still running under a meter).

All of the existing forms of external inputs are for device inputs, or something in the host application giving direction to swingset. We're planning to route all of those through a common protective queue (#720), but this child-process handler is different. The child messages *must* be processed immediately: when one arrives, the kernel ought to be in a state where it's doing nothing except waiting for the message. The kernel has a larger loop of activity, and device inputs must be queued to run not-inside that loop. While the kernel is processing a crank for a child-process vat worker, it is in a nested smaller loop of activity, where the only state is "wait for a child message", and nothing else should happen concurrently.

**Comment by dckc** [comment]
> ... If that's the case, #2928 should fix it.

yes, now that we have a diagnosis, I'm less interested in fixes for this issue by itself.

**Comment by dckc** [comment]
fixed in the course of #2928 etc.
